compiler/testData/codegen/bytecodeText/stringOperations/multipleNestedConcat.kt

```diff
@@ -4,7 +4,7 @@
 // - In argument to 2nd call to foo() inside string literal
 fun test(s1: String, s2: String, s3: String): String {
     fun foo(s: String) = s
-    return "foo: " + foo(s1 + s2 + " ${foo("\${s3.length} = ${s3.length}")}")
+    return "foo: " + foo(s1 + s2 + " ${foo("\${s3.length} = ${s3.length} ")}") + " "
```

Comment:
@sfs Why test update is required?

Reply:
@sfs Oh..seems 'Intrinsics.stringPlus'. It's intresting that old backend uses 'Intrinsics.stringPlus' only if receiver is nullable, with receiver is non-nullable than StringBuilder is used directly. It's not clear which way is better. Is JVM differ these two cases on any string optimizations? //cc @udalov

Reply:
HotSpot optimizes string concatenation (`"a" + "b" + "c"`, which is, effectively, `new StringBuilder().append("a").append("b").append("c")`).

Reply:
@dnpetrov Yes pr does same, 
the question is 
```
new StringBuilder
dup
init
aload var1
append
aload var2 or LDC constant
append 
toString
```
would be same as efficient on Hotspot  and other JVMs as
```
aload var1
aload var2 or LDC constant
invoke Intrinsics.stringPlus

fun Intrinsics.stringPlus(var1, var2)
new StringBuilder
dup
init
aload var1
append
aload var2
append 
toString
```

Reply:
Having a single stringPlus call is slightly less code than creating a new StringBuilder.

If there are no plans to remove the stringPlus intrinsic then we might as well use it. I don't think there would be any performance difference in practice, since Intrinsic.stringPlus is a small leaf function which should be inlined (when hot) and uses StringBuilder.append internally.

Reply:
@sfs Yes it should be inlined, but after inlining it could be a bit different bytecode, so if some pattern based string optimization is used further...

Reply:
@max-kammerer I can't give a definite answer to this, but as far as I can tell the StringBuilder related optimizations (StringOpts) run [as part of inlining](https://github.com/sourcemirror/jdk-8-hotspot/blob/af290d0fb302ee6a355fc55bc11602fddbac08bb/src/share/vm/opto/compile.cpp#L2013).